### DISCUSSION TITLE
Does AvaloniaUI support OpenGL ES 3.1+?

### DISCUSSION THREAD

**Discussion by anose001** [AvaloniaUI/Avalonia#16188] [Q&A] [answered]
I setup an OpenGL control via [example](https://github.com/SamboyCoding/OpenTKAvalonia) and latest Angle version 2.1.2.21 via NuGet package. On my desktop this provides OpenGL ES 3.0. Is OpenGL ES 3.1+ supported via AvaloniaUI?

**Comment by maxkatz6** [comment]
Avalonia uses Angle from October 6 - https://github.com/AvaloniaUI/angle/commits/avalonia-fork-master/
Which has incomplete support for 3.1 - https://github.com/AvaloniaUI/angle/blob/avalonia-fork-master/doc/ES31StatusOnD3D11.md
The same status in on latest master branch as well - https://github.com/google/angle/blob/main/doc/ES31StatusOnD3D11.md

If you need any of missing features in AngleGL, you can use WGL rendering mode. But WGL is known to have poorer stability with some drivers.

**Reply by anose001** [reply]
@maxkatz6, I see. I added the latest Avalonia Angle version via NuGet on my Android project. I run this OpenGL [Avalonia Silk example](https://github.com/kekekeks/Avalonia-Silk.NET-Example/tree/master/AvaloniaSilkExample) with the OpenGLES Silk.NET package.

I checked `_gl.GetStringS(StringName.Version);` which returns **OpenGL ES 3.0.0 (ANGLE 2.1.22083 git hash: a30c67f61af3)** which according to [GitHub ANGLE](https://github.com/google/angle) and your comment should run _OpenGL ES 3.2: ANGLE 2.1.2.21688.59f158c1695f (**Sept, 2023**)_ or at least _OpenGL ES 3.1: ANGLE 2.1.0.f5dace0f1e57 (**Jul, 2020**)_. For some reason Avalonia does not create an OpenGL ES 3.1/3.2 context.

I traced this back to `CreateSharedContext()` in `OpenGlControlBaseResources.TryCreate()` in `OpenGlControlBase`
```csharp
 OpenGlControlBaseResources.TryCreate(CompositionDrawingSurface surface,
        ICompositionGpuInterop interop,
        IOpenGlTextureSharingRenderInterfaceContextFeature feature)
{        
        IGlContext? context;
        try
        {
            context = feature.CreateSharedContext();
        }
        ...
}
```
where `CreateSharedContext()` has an optional argument to specifiy the GlVersion `CreateSharedContext(IEnumerable<GlVersion>? preferredVersions = null);` but in the implementation no `GlVersion` argument is provided. So I assume it defaults back to the value when the `OpenGlControlBase` solution was implemented (OpenGL ES 3.0) even though the latest ANGLE version supports 3.1 or 3.2?

Is there any way to set the OpenGL ES version on the `OpenGlControlBase`? Am I missing anything here? Any help with this is greatly appreciated.

**Reply by maxkatz6** [reply]
In general, default versions are defined by `AngleOptions`.
You can set them in your AppBuilder:
```
.With(new AngleOptions
{
    GlProfiles = ...
})
```

**Reply by maxkatz6** [reply]
But it seems Angle Egl ignores these options right now:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.OpenGL/Egl/EglDisplayUtils.cs#L37-L42

Probably because it somehow needs to be passed down from the Win32 project to OpenGL project.

I am not really familiar with opengl related code, but it seems to be a reason of the problem.
If you are interested, you can try to fix that. But we might need @kekekeks opinion.

**Reply by kekekeks** [reply]
Versions list is passed through EglDisplay ctor - https://github.com/AvaloniaUI/Avalonia/blob/9c6a2c931021aaf8a86da8c55ebd21e0609d2c2a/src/Windows/Avalonia.Win32/OpenGl/Angle/AngleWin32EglDisplay.cs#L41-L46

**Reply by maxkatz6** [reply]
@anose001 so it seems you just need to set GlProfiles properly.

**Reply by anose001** [reply]
@maxkatz6, @kekekeks, thanks for the replies. It seems like `AngleOptions` only allows to set OpenGLES 2 or 3 if I am not mistaken?

https://github.com/AvaloniaUI/Avalonia/blob/56108c1caf51122ca6cdf462b5df2b53cf6abae8/src/Windows/Avalonia.Win32/AngleOptions.cs#L14C1-L18C11

Since the forked ANGLE version supports OpenGL ES 3.1 on Desktop GL (I assume this is OpenGL), OpenGL ES, Vulkan, (and almost fully) Direct3D11 rendering backends, is this as easy as adding `new GlVersion(GlProfileType.OpenGLES, 3, 1)` in the above file, calling it in my AppBuilder, and playing around with the functionalities?

Also, does an Avalonia cross-platform app use Direct3D11 as the backing renderer on windows? What does it use on Android or iOS? I am asking this because my standalone Avalonia application uses OpenGL 4.0 as the highest version and I am trying to understand whether I can change the Avalonia cross-platform rendering backend on Desktop to OpenGL and do OpenGL ES on mobile, and if I can, how so?

**Reply by maxkatz6** [reply, marked as answer]
@anose001 no, you can set values other than 2 or 3, it's a mutable property:
```
.With(new AngleOptions
{
    GlProfiles = new [] { new GlVersion(GlProfileType.OpenGLES, 3, 1) }
})
```

> Also, does an Avalonia cross-platform app use Direct3D11 as the backing renderer on windows?

Yes, on Windows.
But also, Windows supports WGL (native GL) and Vulkan.

> Android, Linux

OpenGL by default, Vulkan is a new option in 11.1.

> iOS, macOS

OpenGL in 11.1 by default, Metal in 11.2 (nightly) by default.

> Browser

WebGL 2 by default in 11.1

**Reply by maxkatz6** [reply]
Each platform options (like, Win32Platform options or AndroidPlatformOptions) have RenderingMode property, which defines what is used and in what priority (that's a list).

**Reply by anose001** [reply]
@maxkatz6, ah great, thank you, I will give this a try and see how it goes.

The recommended approach to use Vulkan on Windows is shown in this [GpuInterop example](https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop), correct?

I see that Vulkan/Metal on mobile and OS is a recent development. Once that is implemented and released will there be an example on how to get started on these platforms?

**Reply by maxkatz6** [reply]
> The recommended approach to use Vulkan on Windows is shown in this [GpuInterop example](https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop), correct?

If you want to access Vulkan interop, then yes. 

> Once that is implemented and released will there be an example on how to get started on these platforms?

What exactly is your question? For developers there won't be any difference. Once default value is changed - it will just work on new renderer. I also plan to document these new platform options when I have time.

But if you are asking about GPU interop, then likely no. This part of the code that requires in-deep knowledge from the developer, and understanding GPU related code before doing anything with Avalonia. And we can only provide samples on how to integrate them.

**Reply by anose001** [reply]
@maxkatz6, maybe I am misunderstanding something but will there be a `VulkanControlBase` or `MetalControlBase` similar to `OpenGlControlBase` for rendering onto a surface/view?

**Reply by maxkatz6** [reply]
@anose001 probably not. As these controls are limiting and have worse performance, comparing to GPU interop APIs. But more convenient, sure.

**Reply by anose001** [reply]
@maxkatz6, is it possible to perform a compute operation on apple platforms (let's say iOS) via Metal framework and then display the result in an Avalonia view via GPU interop without having to copy textures from the GPU to the CPU? 
Can you gauge on whether this is possible with Avalonia right now please? If so, how would I go about this?

**Reply by maxkatz6** [reply]
No, metal GPU interop is not currently supported.

**Reply by anose001** [reply]
Ok, i see, so then, what is the next best option to access the metal framework and plot onto an Avalonia view on iOS? Lets say other than OpenGlControlBase

**Reply by maxkatz6** [reply]
For macOS, the easiest way to do custom render is to disable Metal and enable OpenGL instead. And continue using OpenGlControlBase (or other OpenGL related options).
You will lose some plain Avalonia performance, but gain custom opengl rendering flexibility.

Other than that - use skia or avalonia drawing context apis.

**Reply by anose001** [reply]
@maxkatz6, I have been thinking about this for the past week, and I am afraid that our project may require OpenGL ES 3.1 for compute capabilities which is not supported on iOS nor macOS via OpenGL ES as only version ES 3.0 is implemented via ANGLE. I agree with you that enabling OpenGL and continuing using OpenGlControlBase would have been the best route forward.

What do you think about running Vulkan via MoltenVK on top of Metal and follow the [VulkanDemo in the Avalonia repo](https://github.com/AvaloniaUI/Avalonia/tree/a06ae34732d5c897b9a627cbe13836e0bafb1943/samples/GpuInterop/VulkanDemo) to implement GPU interop?
If the GPU interop in that way is not possible then implementing `Marshal.Copy()` operation and write to a `WriteableBitmap` with the given GPU-CPU data transfer bottleneck may be the only option?

**Reply by kekekeks** [reply]
We currently don't have GPU interop APIs for macos, those are planned for undefined future date. Marshal.Copy might be the best option, yes.

**Reply by anose001** [reply]
Thank you!
